### PR TITLE
fix: it.skip no longer causes hooks to be assigned to the wrong test

### LINKED ISSUE #7874: "Cannot set property 'err' of undefined" error occurs in place of other errors during automatic test-reruns on file save.
### Current behavior:
When running tests locally through `cypress open`, and when interrupting tests by saving files (forcing a re-run), any failures occurred during the run can trigger this message:

```
The following error originated from your application code, not from Cypress.

  > Cannot set property 'err' of undefined

When Cypress detects uncaught errors originating from your application it will automatically fail the current test.

This behavior is configurable, and you can choose to turn this off by listening to the uncaught:exception event. Learn more
```

The error is not really originating from the application; it's occurring within `hookFailed` at https://github.com/cypress-io/cypress/blob/4cfcae28f097e013ffe7dc7419905eccf1022709/packages/driver/src/cypress/runner.js#L526. Both `getTest()` and `getTestFromHookOrFindTest(hook)` end up returning `undefined`.

I've seen this on two very different projects.

* At work, which is a custom server + Webpack setup, webpack/React, running on MacOS.
* At home, on a personal project running Next.js, running on Windows (Powershell). This project is linked below.

### Desired behavior:
The test fails with the normal readable stack trace.

### Test code to reproduce
This is easily reproducible on this project (created a branch with a failing spec):
https://github.com/primitiveconcept/ludumdare46/tree/cypress/err-of-undefined/client

* Run `npm run cy:dev` for server + Cypress
* Run the `filesystem.spec.ts` spec
* Save the `filesystem.spec.ts` while the test is running. Reproduces around 25% of the time, may require a few tries.

If needed, I can try to come up with a minimal reproduction, but this isn't a very complex project. The server isn't required for the tests.

### Versions

Cypress versions: Can reproduce on 4.9.0. Cannot reproduce on 4.5.0. Other versions are unstable for various reasons on the example project. At work, this started occurring on 4.6.0.
OS: MacOS + Windows
Browser: Chrome + Electron

┆Issue is synchronized with this [Jira Bug](https://cypress-io.atlassian.net/browse/TR-160) by [Unito](https://www.unito.io/learn-more)

### LINKED ISSUE #8086: it.skip before describe causing problems in both command line run and debugger
### Current behavior: 
I was using a structure in my test cases similar to this one:

```js
describe("UPPER", () => {

    before(() => {})
    beforeEach(() => {})

    it("TEST1", () => {})
    it.skip("TEST2", () => {})

    describe("LOWER", () => {

        before(() => {})
        beforeEach(() => {})
        
        it("TEST3", () => {})
        it("TEST4", () => {})
    })
})
```

And I noticed that when I run this test suite in debugger it skipped all test cases in "LOWER". Also when I was running it in command line (chrome in headless mode) it cause an error  _"TypeError: Cannot read property 'addCommand' of undefined"_. It was because of skipping that one test case directly before describe (TEST2).

Screenshot from debugger:
image

Error output from running test in headless mode in chrome:

      1) UPPER
           "before all" hook for "TEST2":
         TypeError: Cannot read property 'addCommand' of undefined
    
    Because this error occurred during a `before all` hook we are skipping the remaining tests in the current suite: `LOWER`
          at Test.addCommand (https://local.legito.com/__cypress/runner/cypress_runner.js:184075:12)
          at https://local.legito.com/__cypress/runner/cypress_runner.js:183574:27
          at RunnablesStore._withTest (https://local.legito.com/__cypress/runner/cypress_runner.js:183614:17)
          at RunnablesStore.addLog (https://local.legito.com/__cypress/runner/cypress_runner.js:183573:18)
          at EventEmitter.<anonymous> (https://local.legito.com/__cypress/runner/cypress_runner.js:181908:22)
          at executeAction (https://local.legito.com/__cypress/runner/cypress_runner.js:49202:13962)
          at EventEmitter.n (https://local.legito.com/__cypress/runner/cypress_runner.js:49202:13826)
          at EventEmitter.emit (https://local.legito.com/__cypress/runner/cypress_runner.js:51345:5)
          at $Cypress.<anonymous> (https://local.legito.com/__cypress/runner/cypress_runner.js:197914:19)
          at $Cypress.../driver/node_modules/eventemitter2/lib/eventemitter2.js.EventEmitter.emit (https://local.legito.com/__cypress/runner/cypress_runner.js:84133:19)
          at $Cypress.parent.<computed> [as emit] (https://local.legito.com/__cypress/runner/cypress_runner.js:170536:33)
          at $Cypress.action (https://local.legito.com/__cypress/runner/cypress_runner.js:164113:28)
          at trigger (https://local.legito.com/__cypress/runner/cypress_runner.js:171548:22)
          at triggerLog (https://local.legito.com/__cypress/runner/cypress_runner.js:171554:12)
          at $Cypress.logFn [as log] (https://local.legito.com/__cypress/runner/cypress_runner.js:171623:5)
          at logFn (https://local.legito.com/__cypress/runner/cypress_runner.js:163898:26)You can see from the output that the problem is in the before hook and when I tried to run test without before hook in describe("LOWER") it run correctly. I temporary resolved this problem by commenting the whole test case instead of using skip but that is not a proper way to do it.### Test code to reproduceI have tied it on this code just for test purposes and to not exposing original code:```jsdescribe("UPPER", () => {```before(() => {
    cy.wait(1000);
})
beforeEach(() => {
    cy.wait(1000);
})

```

```

it("TEST1", () => {
    cy.wait(1000);
})
it.skip("TEST2", () => {
    cy.wait(1000);
})

```

```

describe("LOWER", () => {

```

```

```
before(() => {
    cy.wait(1000);
})
beforeEach(() => {
    cy.wait(1000);
})

it("TEST3", () => {
    cy.wait(1000);
})
it("TEST4", () => {
    cy.wait(1000);
})
```

})

```

})
```

### Versions
Cypress: 4.11.0 and also tried on 4.10.0
OS: Windows 10 Pro
Browser: Chrome 84

Also failed by using cypress docker image: cypress/included:4.11.0 and cypress/included:4.10.0

┆Issue is synchronized with this [Jira Bug](https://cypress-io.atlassian.net/browse/TR-164) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by panzarino**
- Closes #8086
- Closes #7874 

### User facing changelog

- Fixes an issue where placing an `it.skip` on the last test before a nested suite with a `before` hook would cause the remainder of tests not to show up
- Prevents error "Cannot set property 'err' of undefined" error occurs in place of other errors during automatic test-reruns on file save. 

### How has the user experience changed?

Given a test that looks like the following:

```javascript
describe("UPPER", () => {
  it.skip("TEST1", () => {})

  describe("LOWER", () => {
    before(() => {
      cy.wait(1000);
    })

    it("TEST2", () => {})
  })
})
```

**`cypress open`**

Before:

Screen Shot 2020-07-29 at 1 02 54 PM

After:

Screen Shot 2020-07-29 at 1 02 10 PM

**`cypress run`**

Before:

Screen Shot 2020-07-29 at 1 06 01 PM

After:

Screen Shot 2020-07-29 at 1 06 51 PM

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub?

### CODE REVIEW

**Comment by sainthkh** [comment]
`Closes #8096` might accidentally close a valid PR.

**Comment by panzarino** [comment]
@sainthkh Thanks for pointing that out, meant to reference `8086`, sorry for the confusion.

File: packages/driver/src/cypress/runner.js [hunk 1]
```diff
@@ -545,7 +545,7 @@ const hookFailed = (hook, err, hookName, getTest, getTestFromHookOrFindTest) =>
   // NOTE: sometimes mocha will fail a hook without having emitted on('hook')
   // event, so this hook might not have currentTest set correctly
   // in which case we need to lookup the test
-  const test = getTest() || getTestFromHookOrFindTest(hook)
+  const test = getTestFromHookOrFindTest(hook)
```
Comment (jennifer-shehane) [inline packages/driver/src/cypress/runner.js:548]: you can remove this `getTest` param from the `hookFailed` call above now.

**Review by jennifer-shehane — COMMENTED** [review]
Manually verified that this fixes the original issue.

**Comment by panzarino** [comment]
Looks like when `.skip` is used (at least in this example), `_runner.on('test', ...` is not being fired, so `setTest()` is never called. This makes me wonder whether we should fix this so that `setTest()` sets the correct test or just remove all usages of `setTest()` and `getTest()`. @Bkucera said he would look into it.

**Comment by Bkucera** [comment]
So the issue was setting hook.currentTest back to `getTest()` when we meant to delete hook.ctx.currentTest and find it from runnables list again, not sure how that wasn't caught. So i undid one of the deletions of `getTest()` and only left necessary changes.

**Review by kuceb — APPROVED** [review]

**Pull request merged**